riptable/rt_str.py

```diff
@@ -915,5 +874,76 @@ def char(self, position: Union[int, List[int], np.ndarray]):
     nb_char_par = _njit_par(_nb_char)
 
 
+MIN_INT = np.iinfo(np.int32).min
+
+
+def _populate_wrappers(cls):
+    """
+    Decorator for the CatString class which populates the string methods and sets the defaults for
+    filtered value fills.
+    """
+    for method, filter_value_fill in [
+        ('upper', None),
+        ('lower', None),
+        ('reverse', None),
+        ('removetrailing', None),
+        ('strlen', MIN_INT),
+        ('index_any_of', MIN_INT),
+        ('index', MIN_INT),
+        ('contains', False),
+        ('startswith', False),
+        ('endswith', False),
+        ('regex_match', False),
+        ('substr', None),
+        ('char', None)
+    ]:
+        setattr(cls, method, cls._build_method(method, filter_value_fill))
+    return cls
+
+
+@_populate_wrappers
+class CatString:
+    """
+    Provides access to FAString methods for Categoricals.
+    All string methods are wrappers of the FAString equivalent with
+    categorical re-expansion and option for how to fill filtered elements.
+    """
+    def __init__(self, cat):
+        from .rt_categorical import CategoryMode
+        self.cat = cat
+        if cat.category_mode == CategoryMode.StringArray:
+            string_list = cat.category_array
+            self.fastring = FAString(string_list)
+        else:
+            raise ValueError(f"Could not use .str in {CategoryMode(cat.category_mode).name}.")
+
+    @classmethod
+    def _build_method(cls, method, filter_value_fill):
+        """
+        General purpose factory for FAString wrappers.
+        """
+        func = getattr(FAString, method)
+        is_property = isinstance(func, (cached_property, property))
+
+        @wraps(func)
+        def wrapper(self, *args, filter_fill_value=filter_value_fill, **kwargs):
+            from .rt_categorical import Categorical
+            if is_property:
+                out = getattr(self.fastring, method)
+            else:
+                out = func(self.fastring, *args, **kwargs)
+            if out.dtype.kind == 'S':
+                out = Categorical(self.cat._fa.copy(), out, base_index=self.cat.base_index)
+                out.category_make_unique(inplace=True)
+                return out
+            else:
+                return where(self.cat.isfiltered(), filter_fill_value, out[self.cat.ikey - 1])
+
+        if is_property:
+            wrapper = property(wrapper)

```

Comment:
Should this be ``wrapper = func(wrapper)`` (or something like that)? I'm wondering if this still does the right thing even if ``func`` is a ``cached_property``.

Reply:
I think this works because the property on CatString just returns the property/cached_property on FAString. So I think caching the property on CatString is redundant.

Reply:
Good question. I think the result is accurate whether the original is a property or cached_property but the additional steps associated with Categorical will be executed each time the property is accessed. 
I guess this could work:
`wrapper = type(func)(wrapper)`

Note we do have unit tests for .strlen for cats, which is a cached_property

Reply:
It's also worth pointing out that a new instance of `CatString` is created every time we access `.str`. So maybe `.str` should be a `cached_property`.